src/main/java/org/truffleruby/core/module/ModuleOperations.java

```diff
@@ -495,68 +496,84 @@ public static InternalMethod lookupMethod(DynamicObject module, String name, Vis
         return method.getVisibility() == visibility ? method : null;
     }
 
-    public static MethodLookupResult lookupSuperMethod(InternalMethod currentMethod, DynamicObject objectMetaClass) {
+    public static MethodLookupResult lookupSuperMethod(InternalMethod currentMethod, DynamicObject objectMetaClass,
+            DeclarationContext declarationContext) {
+
         assert RubyGuards.isRubyModule(objectMetaClass);
         final String name = currentMethod.getSharedMethodInfo().getName(); // use the original name
         return lookupSuperMethod(
                 currentMethod.getDeclaringModule(),
-                null,
+                new SuperMethodLookup(null, false, null),
                 name,
                 objectMetaClass,
-                currentMethod.getDeclarationContext());
+                declarationContext).getResult();
     }
 
+
     @TruffleBoundary
-    private static MethodLookupResult lookupSuperMethod(DynamicObject declaringModule, DynamicObject lookupTo,
+    private static SuperMethodLookup lookupSuperMethod(DynamicObject declaringModule, SuperMethodLookup lookup,
```

Comment:
Do we need an extra object SuperMethodLookup here, or could we just pass more arguments?
The call above only extracts the results, and the recursive calls could pass extra arguments.
We could have a Memo<Boolean> or boolean[1] for `foundDeclaringModule` if it needs to be modified by recursive calls and read in the caller.

I see this method already has many arguments, but if that's the worry maybe we should encapsulate all of them in some arguments object. I tend to prefer explicit arguments because it's clearer what are the inputs.

Reply:
Is the complexity is the reason why you want to change it? It probably should not increase footprint, but maybe affect performace a little bit since we allocate a couple of additional `SuperMethodLookup` objects. 

The reason why I added extra wrapper `SuperMethodLookup` is I don't like\prefer methods that return the result, but at the same time modifying something "hidden" (in our case return `MethodLookupResult` and modify `foundDeclaringModule`). It’s good when one of the two is used (it is advisable to explicitly work with the returned result), but not both at once.

However, `Memo<boolean>` It looks fine to me because it "screams" about its purpose to keep something between calls (I just did not know about this class).

Reply:
My main concern here is readability, I find lookupSuperMethod() very hard to follow with both MethodLookupResult and SuperMethodLookup. One possibility would be to have SuperMethodLookup replaces the MethodLookupResult for lookupSuperMethod() methods, so we'd have only one.
But I think the Memo caries the intention more clearly: it's state only useful for recursion, not for the caller.